include/motor-control/core/tasks/usage_storage_task.hpp

```diff
@@ -1,7 +1,182 @@
+
 static constexpr uint16_t distance_data_usage_len = 8;
 
+/**
+ * The message queue message handler.
+ */
+template <can::message_writer_task::TaskClient CanClient,
+          eeprom::task::TaskClient EEPromClient>
+class UsageStorageTaskHandler : eeprom::accessor::ReadListener {
+  public:
+    UsageStorageTaskHandler(
+        CanClient& can_client, EEPromClient& eeprom_client,
+        eeprom::dev_data::DevDataTailAccessor<EEPromClient>& tail_accessor)
+        : can_client{can_client},
+          usage_data_accessor{eeprom_client, *this, accessor_backing,
+                              tail_accessor} {}
+    UsageStorageTaskHandler(const UsageStorageTaskHandler& c) = delete;
+    UsageStorageTaskHandler(const UsageStorageTaskHandler&& c) = delete;
+    auto operator=(const UsageStorageTaskHandler& c) = delete;
+    auto operator=(const UsageStorageTaskHandler&& c) = delete;
+    ~UsageStorageTaskHandler() final = default;
+
+    void handle_message(const TaskMessage& message) {
+        std::visit([this](auto m) { this->start_handle(m); }, message);
+    }
+
+    void read_complete(uint32_t message_index) final {
+        std::ignore = message_index;
+        std::visit([this](auto m) { this->finish_handle(m); }, buffered_task);
+        ready_for_new_message = true;
+    }
+
+    auto ready() -> bool {
+        return ready_for_new_message && usage_data_accessor.read_write_ready();
+    }
+
+  private:
+    void start_handle(const std::monostate& m) { static_cast<void>(m); }
+
+    void finish_handle(const std::monostate& m) { static_cast<void>(m); }
+
+    void start_handle(const usage_messages::GetUsageRequest& m) {
+        ready_for_new_message = false;
+        buffered_task = m;
+        _ensure_part(m.distance_usage_key, distance_data_usage_len);
+        usage_data_accessor.get_data(m.distance_usage_key, 0);
+    }
+
+    void finish_handle(const usage_messages::GetUsageRequest& m) {
+        uint64_t distance_usage = 0;
+        std::ignore = bit_utils::bytes_to_int(
+            accessor_backing.begin(),
+            accessor_backing.begin() + distance_data_usage_len, distance_usage);
+        can::messages::GetMotorUsageResponse resp = {
+            .message_index = m.message_index,
+            .distance_usage_um = distance_usage};
+        can_client.send_can_message(can::ids::NodeId::host, resp);
+    }
+
+    void start_handle(const usage_messages::IncreaseDistanceUsage& m) {
+        ready_for_new_message = false;
+        buffered_task = m;
+        _ensure_part(m.key, distance_data_usage_len);
+        usage_data_accessor.get_data(m.key, 0);
+    }
+
+    void finish_handle(const usage_messages::IncreaseDistanceUsage& m) {
+        uint64_t old_value = 0;
+        std::ignore = bit_utils::bytes_to_int(
+            accessor_backing.begin(),
+            accessor_backing.begin() + distance_data_usage_len, old_value);
+        if (old_value == 0xFFFFFFFFFFFFFFFF) {
```

Comment:
let's make this a constexpr so it's explicitly typed and we're not relying on type promotion